src/os/bluestore/Allocator.h

```diff
@@ -16,11 +16,186 @@
 #include <ostream>
 #include "include/ceph_assert.h"
 #include "bluestore_types.h"
+#include "common/ceph_mutex.h"
 
 typedef interval_set<uint64_t> release_set_t;
 typedef release_set_t::value_type release_set_entry_t;
 
 class Allocator {
+protected:
+
+  struct ExtentCollectionTraits {
+    size_t num_buckets;
+    size_t base_bits; // min extent size
+    size_t base = 1ull << base_bits;
+    size_t factor;  // single bucket size range to be
+                    // determined as [len, len * factor * 2)
+                    // for log2(len) indexing and
+                    // [len, len + factor * base)
+                    // for linear indexing.
+
+
+    ExtentCollectionTraits(size_t _num_buckets,
+                           size_t _base_bits = 12,  //= 4096 bytes
+                           size_t _factor = 1) :
+      num_buckets(_num_buckets),
+      base_bits(_base_bits),
+      base(1ull << base_bits),
+      factor(_factor)
+    {}
+
+    /*
+     * Determines bucket index for a given extent's length in a bucket collection
+     * with log2(len) indexing.
+     * The last bucket index is returned for lengths above the maximum.
+     */
+    inline size_t _get_p2_size_bucket(uint64_t len) const {
+      size_t idx;
+      ceph_assert(factor);
+      const size_t len_p2_max =
+        base << ((factor * (num_buckets - 2)));
+      if (len <= base) {
+        idx = 0;
+      } else if (len > len_p2_max) {
+        idx = num_buckets - 1;
+      } else {
+        size_t most_bit = cbits(uint64_t(len - 1)) - 1;
+        idx = 1 + ((most_bit - base_bits) / factor);
+      }
+      ceph_assert(idx < num_buckets);
+      return idx;
+    }
+    /*
+     * Determines bucket index for a given extent's length in a bucket collection
+     * with linear (len / min_extent_size) indexing.
+     * The last bucket index is returned for lengths above the maximum.
+     */
+    inline size_t _get_linear_size_bucket(uint64_t len) const {
+      ceph_assert(factor);
+      size_t idx = (len / factor) >> base_bits;
+      idx = idx < num_buckets ? idx : num_buckets - 1;
+      return idx;
+    }
+  };
+
+  /*
+   * Lockless stack implementation
+   * that permits put/get operation exclusively
+   * if no waiting is needed.
+   * Conflicting operations are omitted.
+   */
+  template <class T>
+  class LocklessWillfulStack {
```

Comment:
It is used only for uint64_t type, which is disk location.
If this is a generic concept, it should go to "common" or related.

Reply:
I don't think it's of much value as a generic concept for now - so don't want to move it to common stuff. Getting rid off template and leaving here if you don't mind